signer/kms/aws/signer.go

```diff
@@ -0,0 +1,165 @@
+	types.CustomerMasterKeySpecEccNistP521,
+}
+
+var awsSupportedHashFuncs = []crypto.Hash{
+	crypto.SHA256,
+	crypto.SHA384,
+	crypto.SHA512,
+}
+
+// SignerVerifier is a cryptoutil.SignerVerifier that uses the AWS Key Management Service
+type SignerVerifier struct {
+	reference string
+	client    client
+	hashFunc  crypto.Hash
+}
+
+// LoadSignerVerifier generates signatures using the specified key object in AWS KMS and hash algorithm.
+func LoadSignerVerifier(ctx context.Context, ksp *kms.KMSSignerProvider) (*SignerVerifier, error) {
+	a := &SignerVerifier{
+		reference: ksp.Reference,
+	}
+
+	var err error
+	a.client, err = newAWSClient(ctx, ksp)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, hashFunc := range awsSupportedHashFuncs {
+		if hashFunc == ksp.HashFunc {
+			a.hashFunc = ksp.HashFunc
+		}
+	}
+
+	if a.hashFunc == 0 {
+		return nil, fmt.Errorf("unsupported hash function: %v", ksp.HashFunc)
+	}
+
+	return a, nil
+}
+
+// NOTE: This might be all wrong but setting it like so for now
+//
+// KeyID returns the key identifier for the key used by this signer.
+func (a *SignerVerifier) KeyID() (string, error) {
+	return a.reference, nil
+}
+
+// Sign signs the provided message using AWS KMS. If the message is provided,
+// this method will compute the digest according to the hash function specified
+// when the Signer was created.
+func (a *SignerVerifier) Sign(message io.Reader) ([]byte, error) {
+	var err error
+	ctx := context.Background()
+	var digest []byte
+
+	var signerOpts crypto.SignerOpts
+	signerOpts, err = a.client.getHashFunc(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("getting fetching default hash function: %w", err)
+	}
+
+	hf := signerOpts.HashFunc()
+
+	digest, _, err = cryptoutil.ComputeDigest(message, hf, awsSupportedHashFuncs)
+	if err != nil {
+		return nil, err
+	}
+
+	return a.client.sign(ctx, digest, hf)
+}
+
+// Verifier returns a cryptoutil.Verifier that can be used to verify signatures created by this signer.
+func (a *SignerVerifier) Verifier() (cryptoutil.Verifier, error) {
+	return a, nil
+}
+
+// Bytes returns the bytes of the public key that can be used to verify signatures created by the signer.
+func (a *SignerVerifier) Bytes() ([]byte, error) {
+	ctx := context.Background()
+	p, err := a.client.fetchPublicKey(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return cryptoutil.PublicPemBytes(p)
+}
+
+// Verify verifies the signature for the given message, returning
+// nil if the verification succeeded, and an error message otherwise.
+func (a *SignerVerifier) Verify(message io.Reader, sig []byte) (err error) {
+	ctx := context.Background()
+	var digest []byte
+
+	var signerOpts crypto.SignerOpts
+	signerOpts, err = a.client.getHashFunc(ctx)
+	if err != nil {
+		return fmt.Errorf("getting hash func: %w", err)
+	}
+	hf := signerOpts.HashFunc()
+
+	// we want to verify remotely unless we explicitly ask to verify locally
+	if os.Getenv("WITNESS_AWS_VERIFY_LOCALLY") == "1" {
```

Comment:
Generally I don't think we should be reaching into the env directly here -- instead I think this option should be supplied to the `SignerVerifier` struct at creation with the Option pattern, and from there we can add support for configuring these via the environment.  I have some thoughts on this we should sync on.

Reply:
I have sometimes been burned by env vars that aren't discoverable in documentation but you glean from source code. This seems like a good pattern for intuitive dev ex

Reply:
sure thing. I (like with other comments) am thinking about reducing the flag toil, but in reality i'm in agreement that these env vars can lead to burns.

Reply:
We do need to support configuration through env, but I think we have an opportunity to make it more generic across the board so we don't have to have hard coded flags strewn about